avl_tree/AvlTree.java

```diff
@@ -0,0 +1,265 @@
+		// These is the method to delete node if it exist
+		// Otherwise it throws an exception
+		if (root == null) {
+			throw new NoSuchElementException("AVL Tree is Empty!!!");
+		}
+		if (root.data == data) {
+			NodeAVL temp = root.right;
+			if (root.left == null && root.right == null)
+				root = null;
+			else if (root.left == null)
+				root = root.right;
+			else if (temp == null)
+				root = root.left;
+			else {
+				int dta = 0;
+				if (root.right.left == null) {
+					root.right.left = root.left;
+					root = root.right;
+				} else {
+					dta = transverseLeftmost(temp);
+					root.data = dta;
+				}
+				if (root.right == null)
+					root.hRight = 0;
+				else
+					root.hRight = Math.max(root.right.hLeft, root.right.hLeft);
+				root = isRotate(root);
+			}
+
+		}
+		if (node.right == null && node.left == null) {
+			throw new NoSuchElementException("element you wanna delete not exist");
+		}
+		if (node.right != null && node.right.data == data) {
+			NodeAVL del = node.right;
+			if (del.right == null && del.left == null)
+				node.right = null;
+			else if (del.left == null)
+				node.right = del.right;
+			else if (del.right == null)
+				node.right = del.left;
+			else {
+				NodeAVL temp = del.right;
+				if (temp.left == null)
+					node.right = node.right.right;
+				else
+					del.data = transverseLeftmost(temp);
+				del.hRight = Math.max(del.right.hLeft, del.right.hRight);
+			}
+		} else if (node.left != null && node.left.data == data) {
+			NodeAVL del = node.left;
+			if (del.right == null && del.left == null)
+				node.left = null;
+			else if (del.left == null)
+				node.left = del.right;
+			else if (del.right == null)
+				node.left = del.left;
+			else {
+				NodeAVL temp = del.right;
+				if (temp.left == null)
+					node.left = node.left.right;
+				del.data = transverseLeftmost(temp);
+				del.hRight = Math.max(del.right.hLeft, del.right.hRight);
+			}
+		} else if (node.data > data) {
+			delNode(node.left, data);
+			if (node.left == null)
+				node.hLeft = 0;
+			else
+				node.hLeft = Math.max(node.left.hLeft, node.left.hRight) + 1;
+		} else if (node.data < data) {
+			delNode(node.right, data);
+			if (node.right == null)
+				node.hRight = 0;
+			else
+				node.hRight = Math.max(node.right.hLeft, node.right.hRight) + 1;
+		}
+		node = isRotate(node);
+	}
+
+	public int transverseLeftmost(NodeAVL node) {
+		// These method is special method which comes
+		// in play when we have to delete a node
+		// which have both childeren.
+		if (node.left.left == null) {
+			int data;
+			if (node.left != null)
+				data = node.left.data;
+			else
+				data = node.data;
+			node.left = null;
+			return data;
+		}
+		node = node.left;
+		int data = transverseLeftmost(node);
+		if (node.left == null)
+			node.hLeft = 0;
+		else
+			node.hLeft = Math.max(node.left.hLeft, node.left.hLeft);
+		node = isRotate(node);
+		return data;
+	}
+}
+
+class NodeAVL {
+	public int hLeft;

```

Comment:
@Hsankesara Maybe using `protected` will be the right thing to do here. https://www.tutorialspoint.com/java/java_access_modifiers.htm